src/middlewared/middlewared/plugins/filesystem_/acl_linux.py

```diff
@@ -12,86 +14,112 @@ class FilesystemService(Service, ACLBase):
-            # TODO: plumb in acltool to handle recursive / traverse so that we
-            # don't break mountpoints
-            # self.acltool(data['path'], 'chown', uid, gid, options)
-            if gid == -1:
-                chown = subprocess.run(['chown', '-R', str(uid), data['path']],
-                                       check=False, capture_output=True)
-            elif uid == -1:
-                chown = subprocess.run(['chgrp', '-R', str(gid), data['path']],
-                                       check=False, capture_output=True)
-            else:
-                chown = subprocess.run(['chown', '-R', f'{uid}:{gid}', data['path']],
-                                       check=False, capture_output=True)
+            return
 
-            if chown.returncode != 0:
-                raise CallError(f"Failed to chown [{data['path']}]: "
-                                f"{chown.stderr.decode()}")
-
-            job.set_progress(100, 'Finished changing owner.')
+        job.set_progress(10, f'Recursively changing owner of {data["path"]}.')
+        options['posixacl'] = True
+        self.acltool(data['path'], 'chown', uid, gid, options)
+        job.set_progress(100, 'Finished changing owner.')
 
     def setperm(self, job, data):
         job.set_progress(0, 'Preparing to set permissions.')
         options = data['options']
         mode = data.get('mode', None)
+        verrors = ValidationErrors()
 
-        uid = -1 if data['uid'] is None else data['uid']
-        gid = -1 if data['gid'] is None else data['gid']
+        uid = -1 if data['uid'] is None else data.get('uid', -1)
+        gid = -1 if data['gid'] is None else data.get('gid', -1)
 
-        self._common_perm_path_validate(data['path'])
+        self._common_perm_path_validate("filesystem.setperm",
+                                        data['path'],
+                                        options.get('recursive', False),
+                                        verrors)
 
-        acl_is_trivial = self.middleware.call_sync('filesystem.acl_is_trivial', data['path'])
+        current_acl = self.middleware.call_sync('filesystem.getacl', data['path'])
+        acl_is_trivial = current_acl['trivial']
         if not acl_is_trivial and not options['stripacl']:
-            raise CallError(
-                f'Non-trivial ACL present on [{data["path"]}]. Option "stripacl" required to change permission.',
-                errno.EINVAL
+            verrors.add(
+                'filesystem.setperm.mode',
+                f'Non-trivial ACL present on [{data["path"]}]. '
+                'Option "stripacl" required to change permission.',
             )
 
+        verrors.check()
+        is_nfs4acl = current_acl['acltype'] == 'NFS4'
+
         if mode is not None:
             mode = int(mode, 8)
 
-        stripacl = subprocess.run(['setfacl', '-b', data['path']],
+        setfaclcmd = 'nfs4xdr_setfacl' if is_nfs4acl else 'setfacl'
```

Comment:
is there a difference between commands when stripping permissions?

Reply:
Non-recursive case, different CLI commands because I didn't want to re-implement POSIX1e setfacl. In case of recursive, the tool handles both acl types.